internal/discovery/discovery.go

```diff
@@ -0,0 +1,129 @@
+/*
+Copyright 2023 The Kubernetes Authors All rights reserved.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package discovery
+
+import (
+	"fmt"
+	"time"
+
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/tools/clientcmd"
+
+	"k8s.io/kube-state-metrics/v2/pkg/options"
+)
+
+// NewDiscoveryClientFromOptions creates a new discovery client from the given options.
+func NewDiscoveryClientFromOptions(options *options.Options) *discovery.DiscoveryClient {
+	config, err := clientcmd.BuildConfigFromFlags(options.Apiserver, options.Kubeconfig)
+	if err != nil {
+		panic(err)
+	}
+	discoveryClient := discovery.NewDiscoveryClientForConfigOrDie(config)
+	return discoveryClient
+}
+
+// fetchObjects fetches all the objects that can be listed from the apiserver, every 30 seconds.
+func (r *GVKMaps) fetchObjects(discoveryClient *discovery.DiscoveryClient) {
+	if !r.Heartbeat.IsZero() && time.Since(r.Heartbeat) < 30*time.Second {
+		return
+	}
+	groupList, err := r.Client.ServerGroups()
+	if err != nil {
+		panic(err)
+	}
+	gToVMap := make(map[string][]string)
+	gToKMap := make(map[string][]string)
+	gVKMap := make(map[[2]string][]string)
+	gKVMap := make(map[[2]string][]string)
+	for _, group := range groupList.Groups {
+		gToVMap[group.Name] = []string{}
+		gToKMap[group.Name] = []string{}
+		for _, version := range group.Versions {
+			gToVMap[group.Name] = append(gToVMap[group.Name], version.Version)
+			gVKMap[[2]string{group.Name, version.Version}] = []string{}
+			resources, err := discoveryClient.ServerResourcesForGroupVersion(version.GroupVersion)
+			if err != nil {
+				panic(err)
+			}
+			for _, resource := range resources.APIResources {
+				gToKMap[group.Name] = append(gToKMap[group.Name], resource.Kind)
+				gVKMap[[2]string{group.Name, version.Version}] = append(gVKMap[[2]string{group.Name, version.Version}], resource.Kind)
+				gKVMap[[2]string{group.Name, resource.Kind}] = append(gKVMap[[2]string{group.Name, resource.Kind}], version.Version)
+			}
+		}
+	}
+	r.gToVMap = gToVMap
+	r.gToKMap = gToKMap
+	r.gVKMap = gVKMap
+	r.gKVMap = gKVMap
+	r.Heartbeat = time.Now()
```

Comment:
having `GVKMaps` have a fake clock and using that can make testing easier (aka clock injection).

Reply:
(also some of this stuff looks pretty unit-testable), like `ResolveGVK()`

Reply:
Instead of r.Heartbeat, we could also use https://gobyexample.com/tickers